enhancements/microshift/dummy-gateway-interface-ovn.md

```diff
@@ -0,0 +1,426 @@
+* Host to ClusterIP services backed by overlay pods
+* Host to NodePort services backed by overlay pods
+
+The path for following north-south traffic case will be slightly different:
+
+* Host to CLusterIP services backed by host pods
+* Host to NodePort services backed by host pods
+* External to NodePort services
+
+### Host to ClusterIP Services Backed by Host Pods
+
+The service and endpoint are:
+
+```bash
+[root@rhel92 microshift]# oc get svc nginx-host
+NAME         TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)        AGE
+nginx-host   NodePort    10.43.117.48   <none>        80:30847/TCP   5d21h
+[root@rhel92 microshift]# oc get ep nginx-host
+NAME         ENDPOINTS            AGE
+nginx-host   10.89.0.3:8080       5d21h
+```
+
+The general flow is:
+1. The packet is sent to the OVN gateway router to br-ex with src_ip (10.44.0.1),
+   dest_ip(10.43.117.48).
+
+   ```none
+   10.43.0.0/16 via 169.254.169.4 dev br-ex mtu 1400
+   ```
+
+2. In br-ex,the packet will be SNATed to 169.254.169.2, then sent to the OVN
+   Gateway Router
+
+   ```none
+   cookie=0xdeff105, duration=17970.101s, table=0, n_packets=4222, n_bytes=388436, priority=500,ip,in_port=LOCAL,nw_dst=10.43.0.0/16 actions=ct(commit,table=2,zone=64001,nat(src=169.254.169.2))
+   cookie=0xdeff105, duration=5.359s, table=2, n_packets=4237, n_bytes=394001, actions=mod_dl_dst:d2:a9:ab:af:55:37,output:"patch-br-ex_rhe"
+   ```
+
+3. The loadbalancer in the GR DNAT the packet endpoint 10.89.0.3:8080. The
+   packet be SNATed to the GR interface IP (10.44.0.1).
+
+   ```none
+    7. lr_in_dnat (northd.c:10553): ct.new && !ct.rel && ip4 && reg0 == 10.43.117.48 && tcp && reg9[16..31] == 80, priority 120, uuid fb1a2cf8
+       flags.force_snat_for_lb = 1;
+       ct_lb_mark(backends=10.89.0.3.129:8080; force_snat);
+   ```
+
+4. According to the default route in GR, the packet is send back to br-ex.
+   
+   ```none
+   sh-5.2# ovn-nbctl lr-route-list GR_ovn-control-plane
+   IPv4 Routes
+   Route Table <main>:
+            169.254.169.0/29             169.254.169.4 dst-ip rtoe-GR_ovn-control-plane
+               10.244.0.0/16                100.64.0.1 dst-ip
+                   0.0.0.0/0             169.254.169.4 dst-ip rtoe-GR_ovn-control-plane
+   ```
+
+5. In br-ex, the packet will be SNAT to 169.254.169.1, and send to local host.
+   The packet will hit the following openflow rules:
+   
+   ```none
+   cookie=0xdeff105, duration=48397.416s, table=0, n_packets=6, n_bytes=928, priority=500,ip,in_port="patch-br-ex_rhe",nw_src=10.44.0.1,nw_dst=10.89.0.3 actions=ct(commit,table=4,zone=64001)
+   cookie=0xdeff105, duration=48397.416s, table=4, n_packets=6, n_bytes=928, ip actions=ct(commit,table=3,zone=64002,nat(src=169.254.169.1))
+   cookie=0xdeff105, duration=10.170s, table=3, n_packets=12647, n_bytes=1517745, actions=move:NXM_OF_ETH_DST[]->NXM_OF_ETH_SRC[],mod_dl_dst:9e:0c:05:73:23:54,LOCAL
+   ```
+
+6. The host routing table then forward the traffic to the hostNetwork pod.
```

Comment:
How many extra hops does the change introduce? Just 1?

Reply:
Zero. Since we use local gw mode in Microshift, the traffic will need to go through the host iptables anyway.

Reply:
Excellent, thanks!